### ISSUE TITLE
Update indent rule with sane defaults for JSX

### ISSUE THREAD

**Issue by kaicataldo** [eslint/eslint#8425]
As discussed in https://github.com/eslint/eslint/pull/7618, now that the indent rewrite has landed, we should enhance it with sane defaults for JSX, allowing for users who prefer to use a custom rule like [eslint-plugin-react](https://github.com/yannickcr/eslint-plugin-react) to do so with an `ignoreJSX` option.

__Note__: It already currently checks JSX. The proposal here is to update the rule to make it allow for common JSX styles and fix any bugs we see. Ideally, people won't have to use the `ignoreJSX` option. Actually, maybe we should hold off on that option and wait and see if anyone asks for it? It would be nice if everyone could just use the indent rule, since we support JSX out of the box!

So far, the following enhancements have been discussed:
- Allow for props to be inline or multiline indented one level
- Closing tags should be inline or indented at same level as opening tag
- The first level of indentation for JS expressions inside JSXExpressionContainer nodes needs to be calculated correctly

**Comment by ljharb** [comment]
It would be really nice if the core rule could handle all jsx-related indentation. Is that not an option?

**Comment by kaicataldo** [comment]
@ljharb What's missing from the list I've started above?

**Comment by not-an-aardvark** [comment]
@ljharb Due to the new design of the indent rule, regardless of what we decide, everything will be checked (unless we explicitly decide to ignore something). So yes, we need to figure out good defaults for all JSX syntax.

**Comment by kaicataldo** [comment]
@ljharb Oh yeah, maybe that wasn't clear. It already currently checks JSX syntax - it just seems like it doesn't account for some common indentation styles, so I was suggesting we add the following JSX-specific behavior to enhance the current behavior.

**Comment by ljharb** [comment]
I'm hoping for good configurability; not just for good defaults - meaning, the options that eslint-plugin-react supports, ideally.

**Comment by soda0289** [comment]
If we added indentation level support, like we do for variable declarators and member expressions, I think that would cover both jsx-indent and jsx-indent-props rules.

**Comment by ljharb** [comment]
The more things that can be moved to core, the better - having to configure indentation rules in multiple places is a mess.

**Comment by not-an-aardvark** [comment]
Looking at the `eslint-plugin-react` docs, it seems like the relevant rules are:

* [jsx-indent](https://github.com/yannickcr/eslint-plugin-react/blob/7111a704435e073f9e33e00e5b82ba4812af7803/docs/rules/jsx-indent.md) only accept the indentation size/type as a config option, which the `indent` rule already uses.
* [jsx-indent-props](https://github.com/yannickcr/eslint-plugin-react/blob/7111a704435e073f9e33e00e5b82ba4812af7803/docs/rules/jsx-indent-props.md) only accept the indentation size/type as a config option, which the `indent` rule already uses.
* [jsx-closing-bracket-location](https://github.com/yannickcr/eslint-plugin-react/blob/7111a704435e073f9e33e00e5b82ba4812af7803/docs/rules/jsx-closing-bracket-location.md) also has an option pertaining to the location of a closing tag.

If we have an `ignoreJSX` option, then I think compatibility with `eslint-plugin-react` is a nice-to-have, but I don't think it's necessary to support all of `eslint-plugin-react`'s options because people can still use `eslint-plugin-react` for the custom behavior.

**Comment by ljharb** [comment]
I agree that core doesn't have to support everything; but anything that a separate plugin has needs to be configurable in the core rule so it doesn't conflict.

It sounds like this should not be a problem here, however :-)

**Comment by kaicataldo** [comment]
So at this point, the only unsupported behavior would be the ability to configure the closing bracket location, correct? Does that seem like a reasonable thing to add in, @not-an-aardvark? I feel like it's fine to add that in as an enhancement since none of these rules are tailored specifically for React.

**Comment by not-an-aardvark** [comment]
If possible, I would like to avoid adding options (aside from general `ignoreJSX` option) to the rule. The schema is already extremely complicated, and adding more options will make it more difficult to generalize the schema for arbitrary nodes in the future.

I think we should implement the `line-aligned` option (as the default behavior), because that is most consistent with how the rest of the rule works. For example, the rule doesn't allow indentation like this:

```js
foo([
           bar,
           baz
    ]);
```

Instead, it requires indentation like this:

```js
foo([
  bar,
  baz
]);
```

Implementing `tag-aligned` would be more complicated, because it's basing the closing tag's indentation off of the offset of the start tag rather than its indentation.

**Comment by kaicataldo** [comment]
Thanks for the input. I do still think "Closing tags should be inline or indented at same level as opening tag" from the original comment up top is a reasonable default, and defer to your judgment since you know the rule best!

**Comment by not-an-aardvark** [comment]
So to clarify, do you think we should use `tag-aligned` as the default? (This would be more complicated to implement, but it's still doable.)

**Comment by kaicataldo** [comment]
I think `"line-aligned"` makes the most sense as a default.

**Comment by soda0289** [comment]
I think line-aligned looks more natural
```jsx
var x = <Hello
  firstName="John"
  lastName="Smith"
/>;
```
vs tag-aligned
```jsx
var x = <Hello
  firstName="John"
  lastName="Smith"
        />;
```

**Comment by not-an-aardvark** [comment]
Ok, it seems like we're in agreement then. However, it's worth noting that this is different from the default in `eslint-plugin-react`, which is `tag-aligned`.

**Comment by kaicataldo** [comment]
Agreed! So the default would allow for the following, right?

```jsx
var x = <Hello
  firstName="John"
  lastName="Smith"
/>;

var x = (
  <Hello
    firstName="John"
    lastName="Smith"
  />
);
```

**Comment by not-an-aardvark** [comment]
Yes, that's right.

**Comment by kaicataldo** [comment]
Cool - I think that's the most expected behavior :+1:

**Comment by not-an-aardvark** [comment]
@soda0289 Note that `tag-aligned` would also accept this (if I'm understanding the docs correctly):

```js
var x = <Hello
           firstName="John"
           lastName="Smith"
        />;
```

**Comment by kaicataldo** [comment]
Would it be possible to support all three by default?

**Comment by kaicataldo** [comment]
I assigned to show I'm championing this and so we can accept it. Willing to work on this, but also happy to hand it off to @not-an-aardvark, if you'd prefer that :)

**Comment by soda0289** [comment]
Would this be considered line-aligned
```jsx
var x = <Hello
           firstName="John"
           lastName="Smith"
    />,
    y = <World
           firstName="John"
           lastName="Smith"
   />;
```
or is this line-aligned
```jsx
var x = <Hello
    firstName="John"
    lastName="Smith"
/>,
y = <World
    firstName="John"
    lastName="Smith"
/>;
```
As long as we allow `jsxIgnore` option I can live with it.

**Comment by kaicataldo** [comment]
Interesting question. Seems like it would be the second. Ideally, I think we should allow for the following, if possible:
```js
var x = <Hello
    firstName="John"
    lastName="Smith"
/>,
y = <World
    firstName="John"
    lastName="Smith"
/>;

var x = <Hello
           firstName="John"
           lastName="Smith"
        />,
        y = <World
           firstName="John"
           lastName="Smith"
       />;
```

**Comment by not-an-aardvark** [comment]
@kaicataldo It would be difficult to do that because the rule has a single desired indentation for any given token.

Luckily, it seems like I was wrong in https://github.com/eslint/eslint/issues/8425#issuecomment-292645337 -- `tag-aligned` doesn't accept the syntax that I gave. So I think defaulting to just `line-aligned`  is fine.

**Comment by not-an-aardvark** [comment]
@soda0289 That would be determined by the `VariableDeclarator` option. Both of the examples would be allowed depending on whether the `VariableDeclarator` config was 1 or 0, respectively.

Basically, if you squint a bit and pretend that the JSXElement is an object literal and the attributes are properties, the same thing will be allowed for the element as for the corresponding object literal.

**Comment by kaicataldo** [comment]
Defaulting to `line-aligned` seems like the best course of action to me, then!

**Comment by soda0289** [comment]
@not-an-aardvark That works

**Comment by not-an-aardvark** [comment]
@kaicataldo Re. implementing, it's up to you -- I'm fine with implementing this, but maybe it would be a good idea for other people to become familiar with the rule.

**Comment by kaicataldo** [comment]
Yeah, sounds good!

**Comment by ljharb** [comment]
@kaicataldo to confirm; regardless of the default, would alternatives to `line-aligned` be configurable?

**Comment by kaicataldo** [comment]
@ljharb as of the current proposal, no - see @not-an-aardvark's comment: https://github.com/eslint/eslint/issues/8425#issuecomment-292635852

I'm not against it, though, and it's something that could be added in the future in a semver-minor release if need be.

**Comment by ljharb** [comment]
Why not? Considering there's a clearly demonstrated need for it in the ecosystem, and considering that upon releasing this default you will *guaranteed* get a number of issues filed immediately asking for these things, why not include it in the mvp?

**Comment by not-an-aardvark** [comment]
Again, I don't think we need full compatibility with `eslint-plugin-react`'s options. We're providing reasonable defaults and also allowing an escape hatch with `ignoreJSX` so that users can use plugins for custom behavior if they choose, without requiring the plugins to fully reimplement the rule. 

Right now I'm unconvinced that we need an option for this, and since it would be slightly complex to implement with the way the rule is set up, I think it's better to wait for now. We can always decide to add it later.

**Comment by yannickcr** [comment]
> I don't think we need full compatibility with `eslint-plugin-react`'s options.

Agree with that, missing some too specific options is fine imho. Ideally most users would just go with the eslint's `indent` rule defaults and others who really want a different style will disable JSX linting with the `ignoreJSX` option and use more specialized rules.

**Comment by not-an-aardvark** [comment]
@kaicataldo By the way, the `ignoreJSX` option will probably share a lot of logic with the fix for https://github.com/eslint/eslint/issues/8440. If you're working on this, I think it would be fine to only change the default JSX behavior first without adding `ignoreJSX`, and then someone can add `ignoreJSX` after https://github.com/eslint/eslint/issues/8440 is fixed.

**Comment by kaicataldo** [comment]
Hoping to get to this this weekend!

**Comment by kaicataldo** [comment]
Wasn't able to get to this this weekend but am planning to do this in the next day or two.

**Comment by kaicataldo** [comment]
I started looking at this yesterday and it's taking longer than I expected. Sorry, really though I could get it done before tomorrow :( If anyone else wants to take it on who can get it done in a more timely manner, please feel free (just leave a comment here so I know to stop working on it!)

**Comment by hallaathrad** [comment]
Hey guys :) great job with the updates. The new strict indentation works great.
However, just updated to 4.0.0 and I'm struggling with the props/attributes of my components; cannot seem to get a proper grip of the rule now. I'm trying to set it to "first" but none of the options in the documentation seem to match. Am I missing something?

**Comment by soda0289** [comment]
@hallaathrad 
Currently the new indent rule only supports the following indentation style for multi line props:
```js
<RandomComponent
    className="stuff"
    prop1={moreStuff}
/>
// OR
<RandomComponent className="stuff"
    prop1={moreStuff}
/>
// OR
<RandomComponent className="stuff"
    prop1={moreStuff} />
```

There are plans to allow JSX indentation to be disabled so that other rules/plugins can indent JSX to your liking. If you need JSX to be ignored currently the only way would be to use the legacy indent rule.

**Comment by hallaathrad** [comment]
Thanks for your quick answer, @soda0289. I was hoping one of the options in the documentation would apply something like arguments or parameters to align to "first" (as props, args, etc are not children of the element, so they shouldn't behave/indent the same as children).

However, I cannot even seem to be able to change it the way you describe just for the JSX, without modifying it for the rest of my JS. How can I ignore JSX using legacy? cannot find the specifics to do this in the documentation.

**Comment by soda0289** [comment]
Currently none of the indent object option apply to JSX props. It could be possible to add a "first" option but you would need to open a new issue to discuss that.

The old indent rule, which is now called `indent-legacy`, does not indent JSX. It is the same rule as the one released in v3.19.

**Comment by hallaathrad** [comment]
Thanks again. Will open a new issue on this subject.

**Comment by ljharb** [comment]
You can also use eslint-plugin-react for more sophisticated jsx indentation rules.

**Comment by hallaathrad** [comment]
@ljharb The two relevant rules conflict (eslint's indent vs. react's indent-props).
Ignoring JSX indent on eslint ignores regular JS inlined therein, which jsx doesn't catch, methinks.
Guessing I'll have to run some experiments and see what works. :(
Thanks for your input.
